src/test/kotlin/com/github/avrokotlin/avro4k/endecode/EnDecoder.kt

```diff
@@ -0,0 +1,125 @@
+import io.kotest.matchers.shouldBe
+import io.kotest.mpp.newInstanceNoArgConstructorOrObjectInstance
+import kotlinx.serialization.DeserializationStrategy
+import kotlinx.serialization.KSerializer
+import kotlinx.serialization.SerializationStrategy
+import kotlinx.serialization.serializer
+import org.apache.avro.Schema
+import org.apache.avro.generic.GenericDatumReader
+import org.apache.avro.generic.GenericDatumWriter
+import org.apache.avro.generic.GenericRecord
+import org.apache.avro.io.Decoder
+import org.apache.avro.io.DecoderFactory
+import org.apache.avro.io.Encoder
+import org.apache.avro.io.EncoderFactory
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.InputStream
+
+
+sealed interface EnDecoder {
+    val name: String
+    var avro: Avro
+    fun encodeGenericRecordForComparison(value: GenericRecord, schema: Schema): ByteArray
+
+    fun <T> decode(
+        byteArray: ByteArray,
+        deserializer: DeserializationStrategy<T>,
+        readSchema: Schema,
+        writeSchema: Schema
+    ): T
+
+    fun <T> encode(value: T, serializer: SerializationStrategy<T>, schema: Schema): ByteArray
+}
+
+class AvroLibEnDecoder : EnDecoder {
+    override var avro: Avro = Avro.default
+    override val name: String = "AvroLibrary"
+    override fun <T> encode(value: T, serializer: SerializationStrategy<T>, schema: Schema): ByteArray {
+        val asRecord = avro.toRecord(serializer, schema, value)
+        return encodeGenericRecordForComparison(asRecord, schema)
+    }
+
+    override fun encodeGenericRecordForComparison(value: GenericRecord, schema: Schema): ByteArray {
+        val writer = GenericDatumWriter<GenericRecord>(schema)
+        val byteArrayOutputStream = ByteArrayOutputStream()
+        val encoder = avroLibEncoder(schema, byteArrayOutputStream)
+        writer.write(value, encoder)
+        encoder.flush()
+        return byteArrayOutputStream.toByteArray()
+    }
+
+    override fun <T> decode(
+        byteArray: ByteArray,
+        deserializer: DeserializationStrategy<T>,
+        readSchema: Schema,
+        writeSchema: Schema
+    ): T {
+        val input = ByteArrayInputStream(byteArray)
+        val reader = GenericDatumReader<GenericRecord>(writeSchema, readSchema)
+        val genericData = reader.read(null, avroLibDecoder(writeSchema, input))
+        return avro.fromRecord(deserializer,genericData)
+    }
+
+    fun avroLibEncoder(schema: Schema, outputStream: ByteArrayOutputStream): Encoder =
+        EncoderFactory.get().jsonEncoder(schema, outputStream)
+    
+    fun avroLibDecoder(schema: Schema, inputStream: InputStream) : Decoder = 
+        DecoderFactory.get().jsonDecoder(schema, inputStream)
+}
+
+inline fun <reified T> EnDecoder.testEncodeDecode(
+    value: T,
+    shouldMatch: RecordBuilderForTest,
+    serializer: KSerializer<T> = avro.serializersModule.serializer<T>(),
+    schema: Schema = avro.schema(serializer)
+) {
```

Comment:
I had in mind that I took this problem into account, but I'm not sure anymore because life has gone on since I submitted this PR ;)

After a quick look at the code, I think this should not be a problem because the serialization result will always be compared against the `shouldMatch` parameter, which is the expected result as an Avro (Apache Avro) record. So, the encoded result will be compared against the reference implementation. I think this is the best we can do.
As an additional safeguard, we could introduce snapshot tests, but I had a bad experience with this kind of tests. They can become quite a hassle to properly maintain. 

I put it into a symmetric test clause because sometimes an Encoder/Decoder pair was not properly tested for each test case in both directions.

Reply:
Oh sorry, I did not got it during the first read.

I also feel a bit complicated the `EnDecoder` class, maybe a bit too complex for what it is in real.

Why not just having 3 methods `testEncodeDecode`, `testEncodeIsEqual` and `testDecodeIsEqual` without this `includeForEveryEncoder` and `EnDecoder` and `AvroLibEnDecoder` ? I would say that we don't want multiple implementations inside avro4k to not have a too complex project to maintain 😄 

If you are not taking this in the next days, I'll push a commit to show an idea 🚀